### PR TITLE
Define agent shell command style.

### PR DESCRIPTION
**Pull request by jeschkies**
Summary:
When the Jenkins agent is launched on a Windows node the command differs slightly
from the Linux command. This patch introduces a command style which defaults to Linux.

### CODE REVIEW

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosAgentSpecTemplate/ContainerInfo/config.jelly [hunk 1]
```diff
@@ -15,6 +15,16 @@
         <f:radioBlock name="type" title="${%Mesos}" value="MESOS" inline="true" checked="${instance.type == 'MESOS'}"></f:radioBlock>
     </f:entry>
 
+    <f:block>
```
Comment (jeschkies) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosAgentSpecTemplate/ContainerInfo/config.jelly]: This is somewhat ugly but `optioanlProperty` did not work on a `String`. Any help is welcome.

File: src/main/java/org/jenkinsci/plugins/mesos/api/LaunchCommandBuilder.java [hunk 2]
```diff
@@ -137,13 +137,19 @@ public LaunchPod build() throws MalformedURLException, URISyntaxException {
...
   /** @return the agent shell command for the Mesos task. */
   private String buildCommand() throws MalformedURLException {
-    return String.format(
-        AGENT_COMMAND_FORMAT,
-        this.xmx,
-        this.jvmArgString,
-        this.jnlpArgString,
-        buildJnlpSecret(),
-        buildJnlpUrl());
+    // Check if command is overriden
```
Comment (jeschkies) [inline src/main/java/org/jenkinsci/plugins/mesos/api/LaunchCommandBuilder.java]: Please see `SimpleTaskInfoBuilder` in USI. The `shellCommand` will call 
```scala
      taskInfoBuilder.setCommand(
        Mesos.CommandInfo
          .newBuilder()
          .setShell(true)
          .setValue(shellCommand)
```
this is different from [1.x](https://github.com/jenkinsci/mesos-plugin/blob/1.x/src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java#L1050) where we have
```java
  commandBuilder.setShell(false);
  commandBuilder.setValue(customShell);
``` 
Kai, do you know why we set shell to `false`?

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosAgentSpecTemplate/ContainerInfo/help-overrideShellCommand.html [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<div>
+  Check if you do not want to use the default command to launch the Jenkins agent. This is useful if
+  you want to start the agent in a Windows container:
+  <code>
+    java -DHUDSON_HOME=jenkins -server -Xmx%dm %s -jar ${MESOS_SANDBOX-.}/agent.jar %s %s -jnlpUrl %s";
```
Comment (jeschkies) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosAgentSpecTemplate/ContainerInfo/help-overrideShellCommand.html]: Kai, do you have a proper example? How would a use know the JNLP URL?

File: src/main/java/org/jenkinsci/plugins/mesos/api/LaunchCommandBuilder.java (hunk 2 above)
Comment (timcharper) [reply src/main/java/org/jenkinsci/plugins/mesos/api/LaunchCommandBuilder.java]: shell is set to `true` if we want to pass a shell script to be evaluated by `sh`; `setShell` effectively should've been called `setShellScript(true)`.

**Comment by jeschkies** [comment]
As @kai pointed out I misunderstood the option. We need a new override shell command option.

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosAgentSpecTemplate/ContainerInfo/help-overrideShellCommand.html (hunk 3 above)
Comment (kaiwalyajoshi) [reply src/main/resources/org/jenkinsci/plugins/mesos/MesosAgentSpecTemplate/ContainerInfo/help-overrideShellCommand.html]: The user wouldn't know this and ideally shouldn't configure the JNLP URL, its too much to push on the user and I haven't seen a case where a user requested a custom url.

Right now its derived from here: https://github.com/jenkinsci/mesos-plugin/blob/1.x/src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java#L216-L222

I'm in favour of not exposign this URL unless there's concrete use case from the user to change it.

**Comment by kaiwalyajoshi** [comment]
The shell command option is needed for the [dind wrapper](https://github.com/rollbar/docker/blob/master/hack/dind) which customers typically use for their build jobs.

I've found when and where to use the custom-comand shell to be really confusing and its mechanics to be equally archane. If possible with the 2.0 plugin, we should really disambiguate the scenarios in which this custom comand shell is used.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosAgentSpecTemplate.java [hunk 4]
```diff
@@ -307,6 +315,27 @@ public DescriptorImpl() {
         load();
       }
     }
+
+    /**
```
Comment (timcharper) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosAgentSpecTemplate.java]: Can you help me understand how this relates to the template command? I'm failing to make the connection

File: src/main/java/org/jenkinsci/plugins/mesos/api/LaunchCommandBuilder.java [hunk 5]
```diff
@@ -40,8 +40,15 @@ public LaunchCommandBuilder() {}
...
-  private static final String AGENT_COMMAND_FORMAT =
+  public static enum AgentCommandStyle {
+    Linux,
+    Windows
+  }
+
+  private static final String LINUX_AGENT_COMMAND_TEMPLATE =
       "java -DHUDSON_HOME=jenkins -server -Xmx%dm %s -jar ${MESOS_SANDBOX-.}/agent.jar %s %s -jnlpUrl %s";
+  private static final String WINDOWS_AGENT_COMMAND_TEMPLATE =
+      "java -DHUDSON_HOME=jenkins -server -Xmx%dm %s -jar %MESOS_SANDBOX%/agent.jar %s %s -jnlpUrl %s";
```
Comment (kaiwalyajoshi) [inline src/main/java/org/jenkinsci/plugins/mesos/api/LaunchCommandBuilder.java]:
```suggestion
      "java -DHUDSON_HOME=jenkins -server -Xmx%dm %s -jar %%MESOS_SANDBOX%%/agent.jar %s %s -jnlpUrl %s";
```
The percentage sign wasn't escaped properly, this leads to failures.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosAgentSpecTemplate.java (hunk 4 above)
Comment (jeschkies) [reply src/main/java/org/jenkinsci/plugins/mesos/MesosAgentSpecTemplate.java]: That was a left over from the old diff.

**Pull request merged**